graphql/batch_executor.go

```diff
@@ -0,0 +1,409 @@
+//   error from propagating all the way to the top of the request stack.
+func resolveNonBatchUnitWithCaching(src interface{}, dest *objectWriter, unit *ExecutionUnit) []*ExecutionUnit {
+	var executionUnits []*ExecutionUnit
+	subDestRes, err := reactive.Cache(unit.ctx, getExecutionCacheKey(src, unit.field, unit.selection), func(ctx context.Context) (interface{}, error) {
+		subDest := newObjectWriter(dest, "")
+		executionUnits = resolveNonBatchUnit(ctx, src, subDest, unit)
+		return subDest.res, nil
+	})
+	if err != nil {
+		dest.Fail(err)
+	}
+	dest.Fill(subDestRes)
+	return executionUnits
+}
+
+// getExecutionCacheKey gets the execution cache key for the provided source.
+func getExecutionCacheKey(src interface{}, field *Field, selection *Selection) resolveAndExecuteCacheKey {
+	value := reflect.ValueOf(src)
+	// cache the body of resolve and execute so that if the source doesn't change, we
+	// don't need to recompute
+	key := resolveAndExecuteCacheKey{field: field, source: src, selection: selection}
+	// some types can't be put in a map; for those, use a always different value
+	// as source
+	if value.IsValid() && !value.Type().Comparable() {
+		// TODO: Warn, or somehow prevent using type-system?
+		key.source = new(byte)
+	}
+	return key
+}
+
+// resolveNonBatchUnit resolves a non-batch field in our graphql response graph.
+func resolveNonBatchUnit(ctx context.Context, src interface{}, dest *objectWriter, unit *ExecutionUnit) []*ExecutionUnit {
+	fieldResult, err := unit.field.Resolve(ctx, src, unit.selection.Args, unit.selection.SelectionSet)
+	if err != nil {
+		dest.Fail(err)
+		return nil
+	}
+	subFieldExecUnits, err := unwrapBatchResult(ctx, []interface{}{fieldResult}, unit.field.Type, unit.selection.SelectionSet, []*objectWriter{dest})
+	if err != nil {
+		dest.Fail(err)
+		return nil
+	}
+	return subFieldExecUnits
+}
+
+// unwrapBatchResult traverses the provided sources and fills in result data and
+// returns new execution units that are required to resolve the rest of the
+// query result.
+func unwrapBatchResult(ctx context.Context, sources []interface{}, typ Type, selectionSet *SelectionSet, destinations []*objectWriter) ([]*ExecutionUnit, error) {
+	switch typ := typ.(type) {
+	case *Scalar:
+		return nil, unwrapBatchScalarResult(sources, typ, destinations)
+	case *Enum:
+		return nil, unwrapBatchEnumResult(sources, typ, destinations)
+	case *List:
+		return unwrapBatchListResult(ctx, sources, typ, selectionSet, destinations)
+	case *Union:
+		return unwrapBatchUnionResult(ctx, sources, typ, selectionSet, destinations)
+	case *Object:
+		return unwrapBatchObjectResult(ctx, sources, typ, selectionSet, destinations)
+	case *NonNull:
+		return unwrapBatchResult(ctx, sources, typ.Type, selectionSet, destinations)
+	default:
+		panic(typ)
+	}
+}
+
+// Resolves the scalar type value for all the provided sources.
+func unwrapBatchScalarResult(sources []interface{}, typ *Scalar, destinations []*objectWriter) error {
```

Comment:
executeBatchScalar?

Reply:
or resolveBatchScalar?

Reply:
went with `resolveBatch` and `resolveScalarBatch`, etc